File: duna-core/src/main/java/io/duna/core/inject/LocalServiceBinderModule.java
/*
 * Copyright (c) 2016 Duna Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.core.inject;

import io.duna.core.context.ClasspathScanner;
import io.duna.core.service.LocalServices;

import com.google.inject.AbstractModule;
import com.google.inject.TypeLiteral;
import com.google.inject.UnsafeTypeLiteral;
import com.google.inject.multibindings.MapBinder;

import java.util.List;
import java.util.Objects;
import java.util.logging.Level;
import java.util.logging.LogManager;
import java.util.logging.Logger;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class LocalServiceBinderModule extends AbstractModule {

    private static final Logger logger = LogManager
        .getLogManager()
        .getLogger(LocalServiceBinderModule.class.getName());

    private final ClasspathScanner classpathScanner;

    public LocalServiceBinderModule(ClasspathScanner classpathScanner) {
        this.classpathScanner = classpathScanner;
    }

    @Override
    protected void configure() {
        logger.info(() -> "");
        logger.fine(() -> "");

        MapBinder<Class<?>, Object> localServicesBinder = MapBinder
            .newMapBinder(binder(),
                new TypeLiteral<Class<?>>() {},
                new TypeLiteral<Object>() {},
                LocalServices.class);

        classpathScanner
            .getLocalServices()
            .parallelStream()
            .forEach(contractClassName -> {
                Class<?> contractClass;

                try {
                    contractClass = Class.forName(contractClassName, false, this.getClass().getClassLoader());
                } catch (ClassNotFoundException e) {
                    logger.log(Level.WARNING, e, () -> "");
                    return;
                }

                if (!contractClass.isInterface()) {
                    logger.warning(() -> "Unable to bind " + contractClass.getName() + ". " +
                        "Contracts must be declared as interfaces.");
                    return;
                }

                TypeLiteral<?> contractTypeLiteral = new UnsafeTypeLiteral(contractClass);

                classpathScanner
                    .getImplementationsList()
                    .get(contractClassName)
                    .parallelStream()
                    .map(impl -> {
                        try {
                            return Class.forName(impl.getClassName(), false, this.getClass().getClassLoader());
                        } catch (ClassNotFoundException e) {
                            logger.log(Level.WARNING, e, () -> "");
                            return null;
                        }
                    })
                    .filter(Objects::nonNull)
                    .forEach(serviceClass -> {

                    });
            });


    }
}

File: duna-core/src/main/kotlin/io/duna/core/inject/LocalServiceBinderModule.kt
/*
 * Copyright (c) 2016 Duna Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.core.inject

import com.google.inject.AbstractModule
import com.google.inject.Scopes
import com.google.inject.TypeLiteral
import com.google.inject.UnsafeTypeLiteral
import com.google.inject.multibindings.MapBinder
import com.google.inject.multibindings.Multibinder
import io.duna.core.classpath.ClassPathScanner
import io.duna.core.service.LocalServices
import io.duna.core.util.Services
import java.lang.reflect.Modifier
import java.util.logging.LogManager

/**
 * Binds contracts to their implementations found in the classpath.
 *
 * @author [Carlos Eduardo Melo][cemelo@redime.com.br]
 */
internal object LocalServiceBinderModule : AbstractModule() {

  @JvmStatic
  private val logger = LogManager.getLogManager()
    .getLogger(LocalServiceBinderModule::class.java.name)

  override fun configure() {
    logger.info { "Configuring local services" }
    logger.fine { "Binding local service contracts" }

    val localContracts = ClassPathScanner.getLocalServices()
        .map { Class.forName(it) }

    val localServices = Multibinder
      .newSetBinder(binder(),
        object : TypeLiteral<String>() {},
        LocalServices::class.java)

    localContracts.forEach contractForEach@ { contractClass ->
      if (!contractClass.isInterface && !Modifier.isAbstract(contractClass.modifiers)) {
        logger.warning { "Unable to bind ${contractClass.canonicalName}. " +
          "Contracts must be either an interface or abstract class." }
        return@contractForEach
      }

      val contractTypeLiteral = UnsafeTypeLiteral(contractClass)

      ClassPathScanner
          .getImplementationsInClasspath(contractClass)
          .map { Class.forName(it) }
          .forEach serviceForEach@ { serviceClass ->
            if (serviceClass.isInterface || Modifier.isAbstract(serviceClass.modifiers)) {
              logger.warning { "Unable to bind ${serviceClass.canonicalName}. " +
                "Implementations must be concrete classes." }
              return@serviceForEach
            }

            val qualifier = Services.getQualifier(serviceClass)

            if (qualifier != null) {
              bind(contractTypeLiteral)
                  .annotatedWith(qualifier)
                  .to(serviceClass)
                  .`in`(Scopes.SINGLETON)

              localServices.addBinding()
                .toInstance("duna:${contractClass.name}@${qualifier.javaClass.name}")
            } else {
              bind(contractTypeLiteral)
                  .to(serviceClass)
                  .`in`(Scopes.SINGLETON)

              localServices.addBinding().toInstance("duna:${contractClass.name}")
            }

            logger.fine { "Bound ${contractClass.canonicalName} -> ${serviceClass.canonicalName}" }
          }
    }
  }
}
